README.md
ow_media_capture cookbook
=========================
Manages OpenWach's NodeMediaCapture application

Requirements
------------

#### packages
- `toaster` - ow_media_capture needs toaster to brown your bagel.

Attributes
----------
TODO: List you cookbook attributes here.

e.g.
#### ow_media_capture::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['ow_media_capture']['bacon']</tt></td>
    <td>Boolean</td>
    <td>whether to include bacon</td>
    <td><tt>true</tt></td>
  </tr>
</table>

Usage
-----
#### ow_media_capture::default
TODO: Write usage instructions for each cookbook.

e.g.
Just include `ow_media_capture` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[ow_media_capture]"
  ]
}
```

Contributing
------------
TODO: (optional) If this is a public cookbook, detail the process for contributing. If this is a private cookbook, remove this section.

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
# CHANGELOG for ow_media_capture

This file is used to list changes made in each version of ow_media_capture.

## 0.1.0:

* Initial release of ow_media_capture

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/sync.rb
# Establish ssh wrapper for the git user

ssh_key = Chef::EncryptedDataBagItem.load("ssh", "git")

# Get group name from gid cause this library is "different"
groups_databag_name = node['ow_users']['groups_databag_name']
groups_item_name = node['ow_users']['groups_databag_item_name']
gids_item = data_bag_item(groups_databag_name, groups_item_name)
gids = gids_item["gids"]

git_ssh_wrapper "ow-github" do
  owner node['ow_media_capture']['git_user']
  group gids[node['ow_media_capture']['service_user_gid'].to_s()]
  ssh_key_data ssh_key['id_rsa']
end

# Checkout and Deploy NodeMediaCapture application
# See Chef's deploy resource docs: 
# http://wiki.opscode.com/display/chef/Deploy+Resource
deploy_revision node['ow_media_capture']['app_root'] do
  repository node['ow_media_capture']['git_url']
  revision node['ow_media_capture']['git_rev'] # or "<SHA hash>" or "HEAD" or "TAG_for_1.0" or (subversion) "1234"
  user node['ow_media_capture']['git_user']
  enable_submodules true
  migrate false
  shallow_clone true
  action :deploy # or :rollback
  git_ssh_wrapper node['ow_media_capture']['git_ssh_wrapper']
  scm_provider Chef::Provider::Git # is the default, for svn: Chef::Provider::Subversion

  notifies :restart, "service["+ node['ow_media_capture']['service_name'] +"]"

  before_restart do
    # create default.yaml
    secrets = Chef::EncryptedDataBagItem.load(node['ow_media_capture']['secret_databag_name'], node['ow_media_capture']['secret_item_name'])

    template node['ow_media_capture']['app_root'] + node['ow_media_capture']['config_path'] do
        source "default.yaml.erb"
        variables({
        :incoming_tmp => node['ow_media_capture']['incoming_tmp'],
        :temp_bucket => node['ow_media_capture']['temp_bucket'],
        :temp_reject_bucket => node['ow_media_capture']['temp_reject_bucket'],
        :site_domain => node['ow_media_capture']['site_domain'],
        :port => node['ow_media_capture']['app_port'],
        :tablename => node['ow_media_capture']['couch_table_name'],

        :process_api_scheme => node['ow_media_capture']['process_api_scheme'],
        :process_api_url => node['ow_media_capture']['process_api_url'],

        :django_api_user => secrets['django_api_user'],
        :django_api_password => secrets['django_api_password'],
        :django_api_url => node['ow_media_capture']['api_url'],
        })
    end

    bash "npm install" do
      user node['ow_media_capture']['git_user']
      cwd node['ow_media_capture']['app_root']
      code <<-EOH
      npm install
      EOH
    end

  end

end

attributes/default.rb
#
# Cookbook Name:: ow_media_capture
# Attributes:: default
#
# Copyright 2013, OpenWatch FPC
#
# Licensed under the AGPLv3
#

# Chef repo
default['ow_media_capture']['secret_databag_name'] 		= "ow_media_capture"
default['ow_media_capture']['secret_item_name'] 		= "ow_media_capture"

# SSL
default['ow_media_capture']['ssl_databag_name'] 		= "ssl"
default['ow_media_capture']['ssl_databag_item_name'] 		= "ssl"

# System
default['ow_media_capture']['app_root']      		= "/var/www/NodeMediaCapture"
default['ow_media_capture']['config_path']       	= "/config/default.yaml"
default['ow_media_capture']['git_user']      		= "git"
default['ow_media_capture']['service_user']      	= "media-capture"
default['ow_media_capture']['service_user_gid']     = 500
default['ow_media_capture']['service_name']      	= "ow_media_capture"
default['ow_media_capture']['git_url']      		= "git@github.com:OpenWatch/NodeMediaCapture.git"
default['ow_media_capture']['git_rev']      		= "HEAD"
default['ow_media_capture']['git_ssh_wrapper']   	= "/home/git/.ssh/wrappers/ow-github_deploy_wrapper.sh"
default['ow_media_capture']['log_path']		    	= "/var/log/ow_media_capture.log"
default['ow_media_capture']['run_script']	    	= "run.sh"

# Nginx
default['ow_media_capture']['http_listen_port']     	= 80
default['ow_media_capture']['https_listen_port']     = 443
default['ow_media_capture']['ssl_dir']				= "/srv/ssl/"
default['ow_media_capture']['ssl_cert']     		= "star_openwatch_net.crt"
default['ow_media_capture']['ssl_key']     			= "star_openwatch_net.key"
default['ow_media_capture']['log_dir']     			= "/var/log/ow/"
default['ow_media_capture']['access_log']     		= "nginx_access_media_capture.log"
default['ow_media_capture']['error_log']     		= "nginx_error_media_capture.log"
default['ow_media_capture']['proxy_pass']     		= "http://localhost:5000"

# NodeMediaCapture
default['ow_media_capture']['incoming_tmp']			= "/tmp"
default['ow_media_capture']['temp_bucket']			= "/internment"
default['ow_media_capture']['temp_reject_bucket']	= "/internment-rejected" 
default['ow_media_capture']['site_domain']			= "localhost"
default['ow_media_capture']['app_port']				= 5000
default['ow_media_capture']['couch_table_name']		= "recordings"

# NodeMediaProcess
default['ow_media_capture']['process_api_schema']	= "https://"
default['ow_media_capture']['process_api_url']		= "alpha.openwatch.net/api/"

# Django
default['ow_media_capture']['django_api_schema']		= "https://"
default['ow_media_capture']['django_api_url']		= "alpha.openwatch.net/api"



metadata.rb
name             'ow_media_capture'
maintainer       'OpenWatch FPC'
maintainer_email 'contact@openwatch.net'
license          'All rights reserved'
description      'Installs/Configures ow_media_capture'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '0.1.0'

recipe "ow_media_capture", "Pulls the OpenWatch NodeMediaCapture code, specifies appropriate user permissions to its root, npm installs packages.json, creates default.yaml from the provided template and starts the application as a service"
recipe "ow_media_capture::sync", "Pulls the specified OpenWatch NodeMediaCapture code, npm installs packages.json, creates default.yaml from the provided template and restarts the application service"

%w{ ubuntu }.each do |os|
  supports os
end

depends "nginx"
depends "git_ssh_wrapper"
depends "user", "~> 0.3.0"

# System

attribute "capture/app_root",
  :display_name => "App root directory",
  :description => "Location to checkout application to",
  :default => "/var/www/NodeMediaCapture"

attribute "capture/config_path",
  :display_name => "configuration file path",
  :description => "filepath of the configuration file within app_root",
  :default => "/config/default.yaml"

attribute "capture/git_user",
  :display_name => "Git user",
  :description => "username to run git commands as",
  :default => "git"

attribute "capture/service_user",
  :display_name => "service user",
  :description => "username to run app service as",
  :default => "media-capture"

attribute "capture/service_user_gid",
  :display_name => "service user groupid",
  :description => "group for user running service",
  :default => "500"

attribute "capture/service_name",
  :display_name => "Service name",
  :description => "The name of the service responsible for running the capture app",
  :default => "ow_media_capture"

attribute "capture/git_url",
  :display_name => "Git repository url",
  :description => "URL of the git repository to checkout",
  :default => "git@github.com:OpenWatch/NodeMediaCapture.git"

attribute "capture/git_rev",
  :display_name => "Git revision",
  :description => "Revision to checkout with git. Can be 'HEAD', tag name or a SHA hash",
  :default => "HEAD"

attribute "capture/git_ssh_wrapper",
  :display_name => "Git ssh wrapper location",
  :description => "Output location for shell script to prep git user for connection to OpenWatch's git repo",
  :default => "/tmp/wrap-ssh4git.sh"

attribute "capture/log_path",
  :display_name => "App Log path",
  :description => "An absolute path",
  :default => "/var/log/ow_media_capture.log"

attribute "capture/run_script",
  :display_name => "App run script",
  :description => "Script to have upstart call. A path relative to the app root",
  :default => "run.sh"

# Nginx

attribute "capture/listen_port",
  :display_name => "Public Port",
  :description => "Nginx will listen on this port for this application",
  :default => "80"


# NodeMediaCapture

attribute "capture/incoming_tmp",
  :display_name => "Temporarily Upload Location",
  :description => "Where node-formidable copies files before being processed",
  :default => "/tmp"

attribute "capture/temp_bucket",
  :display_name => "Temporarily Upload Bucket",
  :description => "Where files are transferred for processing",
  :default => "/internment"

attribute "capture/temp_reject_bucket",
  :display_name => "Temporarily Reject Upload Bucket",
  :description => "Where duplicate uploads are transferred for later upload to reject bucket",
  :default => "/internment-rejected"

attribute "capture/site_domain",
  :display_name => "Site domain",
  :description => "",
  :default => "localhost"

attribute "capture/app_port",
  :display_name => "App port",
  :description => "",
  :default => "5000"

attribute "capture/couch_table_name",
  :display_name => "CouchDB tablename",
  :description => "media capture database table name",
  :default => "recordings"

# NodeMediaProcess

  attribute "capture/process_api_schema",
  :display_name => "NodeMediaProcess API Schema",
  :description => "openwatch NodeMediaProcess service scheme",
  :default => "https://"

attribute "capture/process_api_url",
  :display_name => "NodeMediaProcess API url",
  :description => "openwatch NodeMediaProcess service url",
  :default => "alpha.openwatch.net/api/"

# Django

attribute "capture/django_api_schema",
  :display_name => "DJango API Schema",
  :description => "openwatch django service scheme",
  :default => "https://"

attribute "capture/django_api_url",
  :display_name => "Django API url",
  :description => "openwatch django service url",
  :default => "alpha.openwatch.net/api/"

attribute "capture/django_api_user",
  :display_name => "Django API username",
  :description => "openwatch django service api username",
  :default => "test"

attribute "capture/django_api_password",
  :display_name => "Django API password",
  :description => "openwatch django service api user pass",
  :default => "test"
